### ISSUE TITLE
Ports are not binding using docker REST API

### ISSUE THREAD

**Issue by godovdm** [MicrosoftDocs/Virtualization-Documentation#328]
First I would like to sorry for my English, but I have a strange problem using docker containers on TP5 and I don't know where else I can ask a question.
I tried to automate some processes and decided to use docker-py liblrary for it. 

I have two docker hosts - one is Linux and one is Win Server 2016 TP5. On both hosts docker daemons are bound to host IP address to listen connections from outside. Then I wrote script using instructions of the library (shortly I just create and start containers). After I run the script I found that container's port on Windows is not bound to host port.

Using Wireshark I investigated POST requests to the daemon and this is one of them:

```
POST /v1.22/containers/create?name=sqlexpress HTTP/1.1
Host: <docker daemon host>
Accept-Encoding: gzip, deflate
User-Agent: python-requests/2.10.0
Accept: */*
Connection: keep-alive
Content-Type: application/json
Content-Length: 261

{"NetworkDisabled": false, "ExposedPorts": {"1433/tcp": {}}, "Image": "we-tests-registry.rd.ptsecurity.ru/smoke/sql_express:10.0.14300.1000", "AttachStderr": true, "AttachStdin": false, "OpenStdin": false, "Tty": false, "AttachStdout": true, "StdinOnce": false}
```

```
POST /v1.22/containers/c297b7d87e644c40bdf3fa62ebe2ad126182bac4109fe62d033be149c8a20e4c/start HTTP/1.1
Host: <docker daemon host>
Accept-Encoding: gzip, deflate
User-Agent: python-requests/2.10.0
Accept: */*
Connection: keep-alive
Content-Type: application/json
Content-Length: 94

{"NetworkMode": "default", "PortBindings": {"1433/tcp": [{"HostIp": "", "HostPort": "1433"}]}}

```

First one is to create container and second one is to start it. As you can see is second request the dict with PortBindings is sent. And it's confirmed in daemon's logs:
`
time="2016-07-01T06:38:48.849621200-07:00" level=debug msg="form data: {\"NetworkMode\":\"default\",\"PortBindings\":{\"1433/tcp\":[{\"HostIp\":\"\",\"HostPort\":\"1433\"}]}}"`

But if I inspect the created container I can not see any port bindings:

`"PortBindings": null`

If I perform docker run -p 1433:1433 locally from powershell it works just fine. The script does not have problems with Ubuntu host. The same POST requests work well.

Where can be a problem?

Docker version 1.12.0-dev, build 2b97201

**Comment by StefanScherer** [comment]
I suggest to run the `docker run -p 1433:1433 ...` from another machine and check eg. with WireShark the POST requests.

**Comment by godovdm** [comment]
Hello! Just checked request for `run` command. In this case two requests are sent - one for `create` and one for `start`. Network configuration is described in `create` request and `start` request has empty body. I mean that part 

`"PortBindings": {"1433/tcp": [{"HostIp": "", "HostPort": "1433"}]}} `

was sent in create request.

But docker-py library sends information about port bindings in `start` request.

**Comment by StefanScherer** [comment]
ping @friism: docker-py is used in docker-compose, so is this related to some of the docker tools?

**Comment by friism** [comment]
@godovdm can you post how you managed to solve the problem?

I don't think the start request is supposed to contain parameters: https://docs.docker.com/engine/reference/api/docker_remote_api_v1.24/#/start-a-container

The relevant docker-py file is here I believe: https://github.com/docker/docker-py/blob/94f8dd31be3a3f1802f6041a801f01a6e316ba49/docker/api/container.py#L303

**Comment by godovdm** [comment]
I used `host_config` and `port_bindings` described here:
https://github.com/docker/docker-py/blob/master/docs/port-bindings.md

**Comment by friism** [comment]
@godovdm so @shin- that this is probably because the Docker on your Ubuntu system is old and still support accepts host config in the `start` command.

**Comment by godovdm** [comment]
@friism I'm using docker 1.11.2 on my Ubuntu host

**Comment by shin-** [comment]
Hi @godovdm
Regardless of the specific version that is being used, Passing host configuration in `Client.start` has been effectively deprecated for a long time (the Docker Engine started moving away from it in 1.3.0). `docker-py` has maintained legacy support in the spirit of limiting breaking changes, but we emit a [deprecation warning](https://github.com/docker/docker-py/blob/master/docker/api/container.py#L358-L363) whenever it's used outside of its purpose.

In short, the fact that your code works on Ubuntu is the anomaly as far as I'm concerned, and the behavior exhibited on Windows is what I would expect to happen.

If you wish to read more about this issue:
- https://github.com/docker/docker-py/pull/822
- https://github.com/docker/docker-py/issues/750

**Comment by enderb-ms** [comment]
Closing this issue due to inactivity. Please free to open the issue again if there is something we could help with.
